File: leetcode019.py
class ListNode():
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution():
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        ls = 0
        t = head
        while t is not None:
            t = t.next
            ls = ls + 1
        k = ls - n
        if k == 0:
            return head.next
        t = head
        for i in range(k-1):
            t = t.next
        l = t.next
        if l == None:
            return None
        elif l.next== None:
            t.next = None
        else:
            t.next = l.next
        return head

    def printh(self, head):
        while head is not None:
            print(head.val)
            head = head.next

if __name__=="__main__":
    h =head = ListNode(1)

    s = Solution()
    t = s.removeNthFromEnd(h,1)
    s.printh(t)

File: leetcodec++/leetcode019.cpp
//
// Created by 黄敏辉 on 2019-06-28.
//

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode new_head(-1);
        new_head.next = head;

        ListNode *slow = &new_head, *fast = &new_head;

        for (int i = 0 ;i < n; i++){
            fast = fast->next;
        }

        while(fast->next){
            fast = fast->next;
            slow = slow->next;
        }

        ListNode *del = slow->next;

        slow->next = slow->next->next;

        delete del;

        return new_head.next;
    }
};